### PR TITLE
UX: add CSS classes on body tag based on topic statuses.

### PR DESCRIPTION
**Pull request by vinothkannans**
This PR will add CSS classes like `unlisted`, `pinned`, and `unpinned` on the body tag.

### CODE REVIEW

File: app/assets/javascripts/discourse/app/components/discourse-topic.js [hunk 1]
```diff
@@ -23,8 +23,6 @@ export default Component.extend(
       "topic.is_warning",
       "topic.category.read_restricted:read_restricted",
       "topic.deleted:deleted-topic",
-      "topic.categoryClass",
-      "topic.tagClasses",
```
Comment (vinothkannans) [inline app/assets/javascripts/discourse/app/components/discourse-topic.js:27]: Instead we're using the [`add-category-tag-classes`](https://github.com/discourse/discourse/blob/42f6c9b6b9db81e2a6a9ca45e76f067980525f67/app/assets/javascripts/discourse/app/components/add-category-tag-classes.js) component.

File: app/assets/javascripts/discourse/app/components/add-topic-status-classes.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    }
+    if (topic.unpinned) {
+      classes.push(`topic-status-unpinned`);
+    }
+    if (classes.length > 0) {
+      $("body").addClass(classes.join(" "));
+    }
+  },
+
+  @observes("topic.invisible", "topic.pinned", "topic.unpinned")
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/add-topic-status-classes.js:34]: can you use didReceiveAttrs instead ?

File: app/assets/javascripts/discourse/app/components/add-topic-status-classes.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  refreshClass() {
+    scheduleOnce("afterRender", this, this._updateClass);
+  },
+
+  _removeClass() {
+    $("body").removeClass((_, css) =>
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/add-topic-status-classes.js:40]: use vanilla js please

File: app/assets/javascripts/discourse/app/components/add-topic-status-classes.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      classes.push(`topic-status-unlisted`);
+    }
+    if (topic.pinned) {
+      classes.push(`topic-status-pinned`);
+    }
+    if (topic.unpinned) {
+      classes.push(`topic-status-unpinned`);
+    }
+    if (classes.length > 0) {
+      $("body").addClass(classes.join(" "));
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/add-topic-status-classes.js:30]: vanilla js

File: app/assets/javascripts/discourse/app/components/add-topic-status-classes.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
+import Component from "@ember/component";
+import { observes } from "discourse-common/utils/decorators";
+import { scheduleOnce } from "@ember/runloop";
+
+export default Component.extend({
+  didInsertElement() {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/add-topic-status-classes.js:6]: should probably be a tagless component

File: app/assets/javascripts/discourse/app/components/add-topic-status-classes.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    if (this.isDestroying || this.isDestroyed) {
+      return;
+    }
+    const topic = this.topic;
+
+    this._removeClass();
+
+    let classes = [];
+    if (topic.invisible) {
+      classes.push(`topic-status-unlisted`);
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/add-topic-status-classes.js:21]:
```suggestion
      classes.push("topic-status-unlisted");
```

File: app/assets/javascripts/discourse/app/components/add-topic-status-classes.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    }
+    if (topic.pinned) {
+      classes.push(`topic-status-pinned`);
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/add-topic-status-classes.js:24]:
```suggestion
      classes.push("topic-status-pinned");
```

File: app/assets/javascripts/discourse/app/components/add-topic-status-classes.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    }
+    if (topic.unpinned) {
+      classes.push(`topic-status-unpinned`);
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/add-topic-status-classes.js:27]:
```suggestion
      classes.push("topic-status-unpinned");
```

File: app/assets/javascripts/discourse/app/components/add-topic-status-classes.js [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+  didInsertElement() {
+    this._super(...arguments);
+    this.refreshClass();
+  },
+
+  _updateClass() {
+    if (this.isDestroying || this.isDestroyed) {
+      return;
+    }
+    const topic = this.topic;
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/add-topic-status-classes.js:15]: there's not much value in doing this, you can just user `this.topic` everywhere after

File: app/assets/javascripts/discourse/app/components/add-topic-status-classes.js (hunk 5 above)
Comment (jjaffeux) [reply app/assets/javascripts/discourse/app/components/add-topic-status-classes.js:7]: I still see the tagname

Comment (vinothkannans) [reply app/assets/javascripts/discourse/app/components/add-topic-status-classes.js:7]: I'm not seeing any tag name in my dev environment. I removed it [here](https://github.com/discourse/discourse/pull/12729/files#diff-b9a444c238bd9d77d0b9357fd515533b51e3d136e8301cbb73339cb41f0c3a91R5).

**Comment by SamSaffron** [comment]
@jjaffeux is this good to merge?

**Review by jjaffeux — APPROVED** [review]

**Pull request merged**